src/serialization/read/primitive_nested.rs

```diff
@@ -35,30 +35,30 @@ fn compose_array<I: Iterator<Item = u32>, F: Iterator<Item = u32>, G: Iterator<I
     let mut inner = vec![];
 
     assert_eq!(max_rep, 1);
-    assert_eq!(max_def, 3);
-    let mut prev_def = 0;
+    assert!(max_def >= 1);
+    let mut first = true;
     rep_levels
         .into_iter()
         .zip(def_levels.into_iter())
         .for_each(|(rep, def)| {
             match rep {
                 1 => {}
                 0 => {
-                    if prev_def > 1 {
+                    if !first {
                         let old = std::mem::take(&mut inner);
                         outer.push(Some(Array::Int64(old)));
                     }
                 }
                 _ => unreachable!(),
             }
-            match def {
-                3 => inner.push(Some(values.next().unwrap())),
-                2 => inner.push(None),
-                1 => outer.push(Some(Array::Int64(vec![]))),
-                0 => outer.push(None),
+            match max_def - def {
```

Comment:
I'm not fully sure this is correct, should behave the same as before but I think only the repetition levels should influence when a new array starts.